HDDS-13176. containerIds table value format change to proto from string

sumitagrawl opened the pull request:
## What changes were proposed in this pull request?

containerIds Table format is currently <Long, String> where key is containerId and value is state. This format does not allow any further data to be stored in value like replicationIndex or need concatenation to string.

To provide the extension to use value to add various field in compatible manner, String is changed to ContainerCreateInfo, which is having state as one of member.

Later on need bases, further field can be added in proto like replication index.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-13176

## How was this patch tested?

- Pending

swamirishi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/ContainerCreateInfo.java:
```diff
@@ -0,0 +1,141 @@
+      Proto3Codec.get(ContainerProtos.ContainerCreateInfo.getDefaultInstance()),
+      ContainerCreateInfo::getFromProtobuf, ContainerCreateInfo::getProtobuf,
+      ContainerCreateInfo.class);
+  private static final Codec<ContainerCreateInfo> CODEC_OLD_VERSION = new ContainerCreateInfoCodec();
```
We can use DelegatedCodec using StringCodec

swamirishi commented:
@errose28 Can you also please take a look at the upgrade part of the patch

swamirishi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/ContainerCreateInfo.java:
```diff
@@ -0,0 +1,141 @@
+   * For write case, it will create ContainerCreateInfo in proto3 format, but write is allowed only after the
+   * finalization  of feature.
+   */
+  public static class ContainerCreateInfoCodec implements Codec<ContainerCreateInfo> {
```
It would be wrong to blindly write with proto. What if the code was never upgraded then we should write back as a string.

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/upgrade/ContainerTableSchemaFinalizeAction.java:
```diff
@@ -0,0 +1,60 @@
+             containerIdsTable.iterator()) {
+      while (iterator.hasNext()) {
+        Table.KeyValue<ContainerID, ContainerCreateInfo> next = iterator.next();
+        containerIdsTable.put(next.getKey(), next.getValue());
```
done

sumitagrawl commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/upgrade/TestDatanodeUpgradeToContainerIdsTable.java:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  @Test
+  public void testContainerTableAccessBeforeAndAfterUpgrade() throws Exception {
```
added

sumitagrawl commented:
> @errose28 Can you also please take a look at the upgrade part of the patch

@swamirishi As discussed, datanode upgrade flow as below, which is non-rolling including all nodes together,
1. prepare OM node (scm and dn do not have prepare step)
2. replace the software for om, scm , dn
3. restart all components
4. verify the state, Dn implicitly in read only only mode
5. Finalize SCM which in-turn finalize DN, and then finalize OM
Service will start working with new schema.

As such write operation is not there when dn in implicit upgrade phase. But check is also added in write operation as precaution here.

Updated all comments.

errose28 commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutFeature.java:
```diff
@@ -41,7 +41,8 @@ public enum HDDSLayoutFeature implements LayoutFeature {
   HBASE_SUPPORT(8, "Datanode RocksDB Schema Version 3 has an extra table " +
-          "for the last chunk of blocks to support HBase.)");
+          "for the last chunk of blocks to support HBase.)"),
+  CONTAINERID_TABLE_SCHEMA_CHANGE(9, "ContainerID table schema to use value type as proto");
```
I suggest a name like this or similar. Just using "change" will not be unique going forward as we might need to change the schema again.
```suggestion
  WITNESSED_CONTAINER_DB_PROTO_VALUE(9, "ContainerID table schema to use value type as proto");
```

errose28 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/ContainerCreateInfo.java:
```diff
@@ -0,0 +1,161 @@
+      Proto3Codec.get(ContainerProtos.ContainerCreateInfo.getDefaultInstance()),
+      ContainerCreateInfo::getFromProtobuf, ContainerCreateInfo::getProtobuf,
+      ContainerCreateInfo.class);
+  private static final Codec<ContainerCreateInfo> CODEC_OLD_VERSION = new ContainerCreateInfoCodec();
```
We need a more specific name for this than "old version", because we could change this table again and then it is not specific which one the old version is.

errose28 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/upgrade/ContainerTableSchemaFinalizeAction.java:
```diff
@@ -0,0 +1,68 @@
+ * Upgrade Action for DataNode for update the table schema data of containerIds Table.
+ */
+@UpgradeActionHdds(feature = CONTAINERID_TABLE_SCHEMA_CHANGE, component = DATANODE, type = ON_FINALIZE)
+public class ContainerTableSchemaFinalizeAction
```
Do we need to do this? Usually we take the approach of leaving the old data in place and readable instead of trying to reformat on upgrade. I guess the catch here is if the codec reads a value, it won't know whether to treat it as a string or proto.

sumitagrawl replied:
yes, it can not detect the value format. Based on versioning, we are detecting if read proto as string or Proto type. So this is required to be done.

swamirishi reviewed (changes requested):
@sumitagrawl Left a few more review comments.

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/upgrade/ContainerTableSchemaFinalizeAction.java:
```diff
@@ -0,0 +1,68 @@
+          writeNewFormatTable.putWithBatch(batch, next.getKey(), next.getValue());
+        }
+      }
+      metadataStore.getStore().commitBatchOperation(batch);
```
updated the code as discussed

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/ContainerCreateInfo.java:
```diff
@@ -0,0 +1,161 @@
+
+    @Override
+    public boolean supportCodecBuffer() {
+      return CODEC.supportCodecBuffer();
```
code is updated, removed above

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/upgrade/ContainerTableSchemaFinalizeAction.java:
```diff
@@ -0,0 +1,68 @@
+    TypedTable<ContainerID, ContainerCreateInfo> writeNewFormatTable = metadataStore.getStore().getTable(
+            containerIdsTable.getName(), ContainerID.getCodec(), ContainerCreateInfo.getNewCodec());
+    try (BatchOperation batch = metadataStore.getStore().initBatchOperation()) {
+      try (TableIterator<ContainerID, ? extends Table.KeyValue<ContainerID, ContainerCreateInfo>> iterator =
```
updated with single try-with-resource statement. We need close both, and order does not matter.

swamirishi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/WitnessedContainerDBDefinition.java:
```diff
@@ -62,7 +61,7 @@ public String getLocationConfigKey() {
   }
 
-  DBColumnFamilyDefinition<ContainerID, String> getContainerIdsTable() {
+  DBColumnFamilyDefinition<ContainerID, ContainerCreateInfo> getContainerIdsTable() {
```
This constructor can initialize from the map of Definitions.
```suggestion
  DBColumnFamilyDefinition<ContainerID, ContainerCreateInfo> getContainerIdsTable() {
      super(COLUMN_FAMILIES);
  }
```

swamirishi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerSet.java:
```diff
@@ -247,6 +250,16 @@ private boolean removeContainer(long containerId, boolean markMissing, boolean r
     }
   }
 
+  private void deleteContainerTable(long containerId) throws StorageContainerException {
```
Can we rename this method. Maybe deleteFromContainerTable or removeContainerTableEntry()

sumitagrawl replied:
done

swamirishi commented:
> @sumitagrawl once we finalize the approach can you run some tests on how long this reformat will take and any memory impact it might have? Assume a dense 400TB node which would have ~80,000 container entries in this DB.

80000 entries shouldn't be that big of a deal. It should run in a few seconds. Loading the JVM might take longer than the upgrade process itself.

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/WitnessedContainerMetadataStoreImpl.java:
```diff
@@ -64,13 +71,53 @@ private WitnessedContainerMetadataStoreImpl(ConfigurationSource config, boolean
+  /**
+   * this will hold old version tables required during upgrade, and these are initialized based on version only.
+   */
+  public static class PreviousVersionBasedTable {
```
done

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/upgrade/ContainerTableSchemaFinalizeAction.java:
```diff
@@ -0,0 +1,68 @@
+    WitnessedContainerMetadataStore metadataStore = arg.getContainer().getWitnessedContainerMetadataStore();
+    Table<ContainerID, ContainerCreateInfo> previousTable
+        = ((WitnessedContainerMetadataStoreImpl) metadataStore).getPreviousVersionBasedTable().getContainerIdsTable();
+    Table<ContainerID, ContainerCreateInfo> currTable =
```
updated with deleteRange.

swamirishi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerSet.java:
```diff
@@ -67,26 +69,28 @@ public class ContainerSet implements Iterable<Container<?>> {
   private long recoveringTimeout;
-  private final Table<ContainerID, String> containerIdsTable;
+  @Nullable
+  private final WitnessedContainerMetadataStore metadataStore;
```
nit: 
```suggestion
  private final WitnessedContainerMetadataStore containerMetadataStore;
```

swamirishi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerSet.java:
```diff
@@ -67,26 +69,28 @@ public class ContainerSet implements Iterable<Container<?>> {
   private long recoveringTimeout;
-  private final Table<ContainerID, String> containerIdsTable;
+  @Nullable
+  private final WitnessedContainerMetadataStore metadataStore;
```
Instead of passing WitnessedContainerMetadataStore we can pass an implementation of a Table interface which would be writing to a particular table WitnessedContainerMetadataStore based on what witnessedStore updates. We can completely abstract this out so that the downstream users don't have to understand the logic.

sumitagrawl replied:
Raised above Jira for this handling is have real value, I do not aggree to have table being cached for long-lived object as it may cause refresh after upgrade.

swamirishi commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/WitnessedContainerMetadataStoreImpl.java:
```diff
@@ -67,13 +74,53 @@ private WitnessedContainerMetadataStoreImpl(ConfigurationSource config, boolean
-  public Table<ContainerID, String> getContainerIdsTable() {
-    return containerIdsTable;
+  public Table<ContainerID, ContainerCreateInfo> getContainerCreateInfoTable() {
+    if (!VersionedDatanodeFeatures.isFinalized(HDDSLayoutFeature.WITNESSED_CONTAINER_DB_PROTO_VALUE)) {
```
Let us create a wrapper Table implementation which would automatically get updated to the finalized table once the layout upgrade is finalized. We can return the wrapped Table interface which would mean the caller of this method can keep this value in memory and we wouldn't have to update anything on the caller behaviour. Right now the caller is having to call witnessedMetadataStore.getContainerCreateInfoTable().put or delete everytime we can completely get this abstracted out.

swamirishi replied:
We could have this Wrapped object present as member in the class itself. So on finalize all we need to update is a boolean value for the present in the wrapped table.

swamirishi replied:
We have multiple instances in the code base where we cache the reference to the table. The expectation from what I understand is that once the metadata store is initialized the caller is free to cache the reference right now this model doesn't prevent that. We should either come up with a model which prevents caching the reference altogether which would mean redesigning the interface or we need to make the implementation abstracted so that it is fool proof to reference caching.

swamirishi replied:
A few instances...
https://github.com/apache/ozone/blob/5d1b43d44fa435b5a304bd1d99e8fa4a60d092cf/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ListIterator.java#L300-L302
https://github.com/apache/ozone/blob/48c985f8f65c9c0d99c2b3fadc74683f05c1efc7/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/QuotaRepairTask.java#L336-L339
https://github.com/apache/ozone/blob/48c985f8f65c9c0d99c2b3fadc74683f05c1efc7/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateManagerImpl.java#L60-L65
Our entire code base is spread with such use cases.

sumitagrawl replied:
It can be taken up with the JIRA as mentioned above for designing and defining interface

swamirishi reviewed (changes requested):
@sumitagrawl I have a better idea to implement the changed containerIds table right now we are touching multiple classes unnecessarily.
